File: ghidra_export.py
# Ghidra Headless Export Script (Python/Jython)
# @author Antigravity
# @category Export

import json
import sys
import os
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

def run():
    print("[*] Starting Ghidra analysis export...")
    if currentProgram is None:
        print("[!] No current program loaded.")
        sys.exit(1)
        
    program_name = currentProgram.getName()
    print("[*] Program Name: {}".format(program_name))
    
    # Initialize Decompiler
    decomp = DecompInterface()
    if not decomp.openProgram(currentProgram):
        print("[!] Failed to open program in decompiler.")
        sys.exit(1)
        
    monitor = ConsoleTaskMonitor()
    fm = currentProgram.getFunctionManager()
    listing = currentProgram.getListing()
    
    functions_data = {}
    
    # Get all functions
    funcs = fm.getFunctions(True) # True means forward order
    for func in funcs:
        func_name = func.getName()
        entry_point = func.getEntryPoint().toString()
        print("[*] Decompiling {} @ {}".format(func_name, entry_point))
        
        # Decompile function
        decomp_results = decomp.decompileFunction(func, 30, monitor)
        decompiled_c = ""
        if decomp_results and decomp_results.decompileCompleted():
            decompiled_f = decomp_results.getDecompiledFunction()
            if decompiled_f:
                decompiled_c = decompiled_f.getC()
        else:
            decompiled_c = "/* Decompilation failed or timed out */"
            
        # Get Disassembly
        disasm_lines = []
        insts = listing.getInstructions(func.getBody(), True)
        for inst in insts:
            addr = inst.getAddress().toString()
            mnemonic = inst.getMnemonicString()
            operands = inst.getDefaultOperandRepresentationList()
            ops_str = ", ".join([str(op) for op in operands]) if operands else ""
            disasm_lines.append("{} {} {}".format(addr, mnemonic, ops_str))
            
        functions_data[func_name] = {
            "name": func_name,
            "entry": entry_point,
            "decompiled": decompiled_c,
            "disassembly": "\n".join(disasm_lines)
        }
        
    # Build complete report
    report = {
        "program_name": program_name,
        "functions": functions_data
    }
    
    # Dynamically resolve script's directory for output path
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_path = os.path.join(script_dir, "ghidra_analysis.json")
    
    with open(output_path, "w") as f:
        json.dump(report, f, indent=4)
        
    print("[*] Ghidra export completed! Written to: {}".format(output_path))

if __name__ == "__main__":
    run()

File: GhidraExport.java
// Ghidra Headless Export Script
// @author Antigravity
// @category Export

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.FileWriter;
import java.io.File;

public class GhidraExport extends GhidraScript {
    @Override
    public void run() throws Exception {
        println("[*] Starting Ghidra Java analysis export...");
        if (currentProgram == null) {
            println("[!] No current program loaded.");
            return;
        }

        DecompInterface decomp = new DecompInterface();
        if (!decomp.openProgram(currentProgram)) {
            println("[!] Failed to open program in decompiler.");
            return;
        }

        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        sb.append("  \"program_name\": \"").append(currentProgram.getName()).append("\",\n");
        sb.append("  \"functions\": {\n");

        FunctionIterator funcs = currentProgram.getFunctionManager().getFunctions(true);
        boolean firstFunc = true;
        while (funcs.hasNext() && !monitor.isCancelled()) {
            Function func = funcs.next();
            String funcName = func.getName();
            String entryPoint = func.getEntryPoint().toString();
            println("[*] Decompiling " + funcName + " @ " + entryPoint);

            DecompileResults results = decomp.decompileFunction(func, 30, new ConsoleTaskMonitor());
            String decompiledC = "";
            if (results != null && results.decompileCompleted()) {
                decompiledC = results.getDecompiledFunction().getC();
            } else {
                decompiledC = "/* Decompilation failed or timed out */";
            }

            // Get Disassembly
            StringBuilder disasm = new StringBuilder();
            InstructionIterator insts = currentProgram.getListing().getInstructions(func.getBody(), true);
            while (insts.hasNext()) {
                Instruction inst = insts.next();
                String addr = inst.getAddress().toString();
                disasm.append(addr).append(" ").append(inst.toString()).append("\n");
            }

            if (!firstFunc) {
                sb.append(",\n");
            }
            firstFunc = false;

            sb.append("    \"").append(funcName).append("\": {\n");
            sb.append("      \"name\": \"").append(funcName).append("\",\n");
            sb.append("      \"entry\": \"").append(entryPoint).append("\",\n");
            sb.append("      \"decompiled\": ").append(escapeJson(decompiledC)).append(",\n");
            sb.append("      \"disassembly\": ").append(escapeJson(disasm.toString())).append("\n");
            sb.append("    }");
        }

        sb.append("\n  }\n");
        sb.append("}\n");

        // Dynamically get script directory to avoid hardcoded paths
        File scriptFile = getSourceFile();
        String scriptDir = scriptFile.getParent();
        String outputPath = scriptDir + File.separator + "ghidra_analysis.json";

        FileWriter writer = new FileWriter(new File(outputPath));
        writer.write(sb.toString());
        writer.close();

        println("[*] Ghidra export completed! Written to: " + outputPath);
    }

    private String escapeJson(String str) {
        if (str == null) return "null";
        StringBuilder sb = new StringBuilder();
        sb.append("\"");
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            switch (c) {
                case '\\': sb.append("\\\\"); break;
                case '"': sb.append("\\\""); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < ' ') {
                        String t = "000" + Integer.toHexString(c);
                        sb.append("\\u").append(t.substring(t.length() - 4));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append("\"");
        return sb.toString();
    }
}
